src/crimson/net/ProtocolV2.cc

```diff
@@ -1928,6 +1971,23 @@ seastar::future<> ProtocolV2::read_message(utime_t throttle_stamp)
 
     // TODO: change MessageRef with seastar::shared_ptr
     auto msg_ref = MessageRef{message, false};
+    uint64_t rand =
+      ceph::util::generate_random_number<uint64_t>(1, RAND_MAX);
+    if (rand % 10000 <
+        local_conf()->ms_inject_delay_probability * 10000.0) {
+      uint64_t delay_period =
+	local_conf()->ms_inject_delay_max * (double)(rand % 10000) / 10000.0;
+      if(delay_period) {
+        logger().warn("{} <== #{} ms_dispatch will delay after {}s",
+                      conn,
+                      message->get_seq(),
+                      delay_period);
+        return seastar::sleep(
+          std::chrono::seconds(delay_period)).then([=] {
+            return dispatchers.ms_dispatch(conn_ref, std::move(msg_ref));
+        });
+      }
+    }
```

Comment:
I think it is not correct to add delay here. Unlike async msgr, crimson msgr always do fast dispatch, which means there is no queue between the connection and dispatcher that is possible to introduce any delay.